src/chunk_insert_state.c

```diff
@@ -500,6 +501,21 @@ chunk_insert_state_destroy(ChunkInsertState *state)
 	ExecCloseIndices(state->result_relation_info);
 	heap_close(state->rel, NoLock);
 
+	/*
+	 * Postgres stores cached row types from `get_cached_rowtype` in the
+	 * contraint expression and tries to free this type via a callback from
+	 * the `per_tuple_exprcontext`. Since we create constraint expressions
+	 * within the chunk insert state memory context, we need to ensure that
+	 * these callbacks are not called once the chunk insert state is freed as
+	 * doing so would read freed memory. At the point which we're freeing
+	 * memory we should be between tuples, so we free the
+	 * `per_tuple_exprcontext`, calling all callbacks, here.
+	 */
+	if (state->estate->es_per_tuple_exprcontext != NULL)
+		FreeExprContext(state->estate->es_per_tuple_exprcontext, true);
```

Comment:
I assume that you've checked its safe to delete this here? Will PG check `NULL` on `estate->es_per_tuple_exprcontext`? Will it recreate it in that case?

What is the difference to doing `ResetPerTupleExprContext()`?

Reply:
Maybe also clarify the above in comments?

Reply:
When it acesses the expression context directly, postgres checks to ensure that the context exists. We have no way of knowing if postgres stores pointers into the expression context’s menory context, though @cevian is convinced it’s safe.

Reply:
IIRC ResetTupleExprContext only clears the memory, it does not call the callbacks, and there’s a pointer in the callbacks to our memory.

Reply:
IIRC ResetTupleExprContext only clears the memory, it does not call the callbacks, and there’s a pointer in the callbacks to our memory.